### PR TITLE
Add a `--compile-time-deps` build flag to `cargo build`

### PR DESCRIPTION
**Pull request by Veykril**
[Rendered](https://github.com/Veykril/rfcs/blob/compile-time-deps/text/0000-cargo-compile-time-deps.md)

### CODE REVIEW

**Comment by Veykril** [comment]
cc @rust-lang/wg-rls-2 @Undin @vlad20012

**Comment by riking** [comment]
Which of the following spellings should be preferred: `cargo test --compile-time-deps` or `cargo build --cfg=test --compile-time-deps` ? If the latter, I think it's clear that this is a flag that only makes sense on `build`.

**Comment by Veykril** [comment]
`cargo test` builds and executes tests which isn't really relevant to the flag so yes if this ends up as a flag it would only really make sense on the `build` command

**Comment by matklad** [comment]
To build the tests today, the appropriate command is `cargo test —-no-run`. As we can have a proc-macro which is in `dev-dependencies`, combining `no-run` with `compile-tile-deps` seems like a reasonable thing to do for an IDE to expand test-specific macros.

In practice though, `build —all-targets` should cover this case I think.

**Comment by Aaron1011** [comment]
This will also be useful when builder Docker images - `cargo build --compile-time-deps` could be used to cache dependencies in a separate layer, without resorting to workaround like https://stackoverflow.com/questions/58473606/cache-rust-dependencies-with-docker-build

**Comment by bjorn3** [comment]
For Docker caching you did want to build all dependencies outside of the workspace, right? Not just build dependencies, proc macros and run build scripts for any package including those in the workspace as this RFC proposes AFAICT.

**Comment by epage** [comment]
For docker, see rust-lang/cargo#2644, particularly [the summary](https://hackmd.io/@kobzol/S17NS71bh), for why this or most solutions will not work.

File: text/0000-cargo-compile-time-deps.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+This adds a specialized flag to cargo, whose main purpose would be in use for IDEs.
+The scope of this flag is somewhat unclear, what exactly fits the "compile time dependency" terminology, what does not.
+It might be that in the future, cargo introduces a feature that IDEs have to somehow handle to continue to be able to fully analyze source code, but that does not fit into this flag which might need to a new command or flag to succeed this one, although the author believes this to be very unlikely.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The current alternatives are the `RUSTC_WRAPPER` or a plain `cargo check` invocation as has been outlined in the motivation section.
+- The `RUSTC_WRAPPER` mostly works today, but this is not necessarily its intended usage, and it is unknown what other problems such a wrapper unconditionally succeeding invoked commands could cause next.
+- The plain `cargo check` is not a decent option for IDEs, as the IDE loses the ability to discern between build-script problems and workspace diagnostics.
```
Comment (epage) [inline text/0000-cargo-compile-time-deps.md:67]: Do the messages, combined with knowing what a workspace member is, convey enough information that these can instead be filtered?

File: text/0000-cargo-compile-time-deps.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+[guide-level-explanation]: #guide-level-explanation
+
+Running `cargo build --compile-time-deps` unlike like `cargo build`, will only build and run artifacts that are relevant for the host platform.
+This encompasses building proc-macros the workspace depends on as well as building and running build-scripts of the packages.
+Future invocations of `cargo build` will not need to re-build and re-run these artifacts if they were built with the same environment.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Add a flag `--compile-time-deps` to `cargo build` that will cause cargo to only build proc-macros and build-scripts as well as running the build-scripts.
```
Comment (epage) [inline text/0000-cargo-compile-time-deps.md:41]: I take it this means that it would exclusively build proc macro direct dependencies and build scripts for root crates (user-selected crates, implicitly or explicitly), and all of the dependencies necessary for doing that?

Would be good to be more explicit

File: text/0000-cargo-compile-time-deps.md (hunk 1 above)
Comment (Veykril) [reply text/0000-cargo-compile-time-deps.md:67]: Whether something is a workspace member or not is irrelevant, we'd need to be able to know whether a diagnostic comes from building a build-script / proc-macro crate. We'd also still need to rely on the ~~unstable~~ [keep-going](https://doc.rust-lang.org/cargo/reference/unstable.html#keep-going) flag. Turns out that one is stable now, nice.

So assuming we can do that filtering, that would be an option, though this would still be slower than the current wrapper approach / a future way to only build / run the build dependencies.

File: text/0000-cargo-compile-time-deps.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+This adds a specialized flag to cargo, whose main purpose would be in use for IDEs.
+The scope of this flag is somewhat unclear, what exactly fits the "compile time dependency" terminology, what does not.
+It might be that in the future, cargo introduces a feature that IDEs have to somehow handle to continue to be able to fully analyze source code, but that does not fit into this flag which might need to a new command or flag to succeed this one, although the author believes this to be very unlikely.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The current alternatives are the `RUSTC_WRAPPER` or a plain `cargo check` invocation as has been outlined in the motivation section.
+- The `RUSTC_WRAPPER` mostly works today, but this is not necessarily its intended usage, and it is unknown what other problems such a wrapper unconditionally succeeding invoked commands could cause next.
```
Comment (RalfJung) [inline text/0000-cargo-compile-time-deps.md:66]: Here's what it causes next: build cache corruption.
https://github.com/rust-lang/cargo/issues/13485

File: text/0000-cargo-compile-time-deps.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+Aside from building and running these things, this is also needed for the IDE to figure out things like a package's `OUT_DIR` location, or the file path of a build proc-macro which is read from cargo's metadata output.
+
+IDEs can already do this with `cargo` as is, by just running `cargo check` as this builds proc-macros and *runs* build scripts already, but this also checks the rest of the project which for this step depending on the project in question may be quite costly using up time and resources for something the IDE is not even interested in at this point.
+This also has the additional downside that this pollutes the metadata output with compiler diagnostics of the project, so if the project fails to build, but the proc-macros and build scripts work fine, the IDE now sees an error in this step where none should be, possibly unnecessarily poking the user about this step failing even though it technically didn't.
```
Comment (epage) [inline text/0000-cargo-compile-time-deps.md:18]: If you are using the json output, couldn't you determine that the error is not relevant and ignore it?

**Comment by epage** [comment]
We discussed this in today's cargo meeting

**Would r-a be able to use this as a never-to-stabilize option?**  It feels like there is another feature hinted at by this need and it would help to accumulate use cases over time in dealing with package / built-target selection to better understand how to serve that need.  We don't want to block the improvement of IDEs until that happens though.

We are sympathetic to the use case and would like to see a solution.  This has the benefit of not boxing us into a corner with the build graph because of how specialized it is.  However, that specialization is its own hurdle when it comes to the user experience.
- As we add features to cargo, they lower the value of all other features.  Each feature makes it harder to be aware of other features, making it so people are less likely to use any of them.   The more specialized a feature is, the less pay off we get for adding it and is more likely there will be additional use cases it won't satisfy, requiring us to add more
- How do we explain this to users?
  - This is three things in one: package selection (direct dep proc-macros), package selection + build target selection (build scripts of workspace members), running of build scripts of workspace members.   One indication of this problem is the question "what help heading does this belong under?".
  - When a user sees this in the help, how are they suppose to know this is relevant to them?

**Comment by lnicola** [comment]
Yes, we already use a number of unstable features (i.e. `RUSTC_BOOTSTRAP`, `__CARGO_TEST_CHANNEL_OVERRIDE_DO_NOT_USE_THIS`), so it shouldn't be a problem if it doesn't get stabilized in the medium (or even long) term.

**Comment by epage** [comment]
In that case and if @Veykril, as the RFC author, is on board, then we can proceed with this as an issue in the Cargo repo, rather than an RFC.  Assumes its marked as "accepted" (I'll mark it as such as soon as someone creates it).

**Comment by vlad20012** [comment]
Important thing is that there should be a way to use this unstable feature on a stable toolchain without RUSTC_BOOTSTRAP=1

**Comment by RalfJung** [comment]
That's not how unstable feature work?

**Comment by bjorn3** [comment]
> I guess one concern is how it might interact with feature detection done by build scripts.

For that rust-analyzer could use `-Zallow-feature=` I think when `RUSTC_BOOTSTRAP` isn't set already. Not every build script handles this fine, but those that don't are broken anyway.

**Comment by Veykril** [comment]
> > I guess one concern is how it might interact with feature detection done by build scripts.
> 
> For that rust-analyzer could use `RUSTFLAGS="-Zallow-feature="` I think when `RUSTC_BOOTSTRAP` isn't set already. Not every build script handles this fine, but those that don't are broken anyway.

Won't that cause rebuilds?

**Comment by bjorn3** [comment]
> Won't that cause rebuilds?

Right, it does.

> Also that itself is a Z flag won't that still need a nightly channel?

If you set `RUSTC_BOOTSTRAP=1` it works on stable and causes `#![feature]` to be disabled entirely even when using `RUSTC_BOOTSTRAP` or nightly.

File: text/0000-cargo-compile-time-deps.md (hunk 4 above)
Comment (Veykril) [reply text/0000-cargo-compile-time-deps.md:18]: Kind of? We'd need to walk the reverse dependency tree for each diagnostic to check if its part of a build dependency to ignore it. I think. As we do want to report if the build script fails to compile (and hence run)

**Comment by epage** [comment]
Cargo checks `RUSTC_BOOTSTRAP`

**Comment by Veykril** [comment]
Opened https://github.com/rust-lang/cargo/issues/14434

**Comment by vlad20012** [comment]
Some build scripts emit cargo::rerun-if-changed=RUSTC_BOOTSTRAP. Namely, this does `anyhow`, that is a dependency of proc_macro2, that is a dependency of basically any non-trivial rust project.

If rust-analyzer invokes cargo with RUST_BOOTSTRAP=1 in a project with such a build script, the project will be rebuilt almost from scratch during the next cargo build invocation without RUSTC_BOOTSTRAP=1.

Basically, a project would be rebuilt from scratch after opening in rust-analyzer.

This is why requiring RUSTC_BOOTSTRAP=1for this feature is very unsatisfactory.

May we have another hacky variable specially for thus use case?
